### PR TITLE
net.stabyourself.nottetris2

### PR DESCRIPTION
**Pull request by Yajo**
Hi there, as an exercise to know flatpak, I'm trying to package [Not Tetris 2](http://stabyourself.net/nottetris2).

This is a simple game that requires an old version of the LÖVE engine to work, so it is a very good candidate as the kind of app that would become a nightmare to package without something like Flatpak.

Besides, it's crazy funny.

This PR is a WIP. I didn't have the intention to open it yet, but it turned out I hit [this error](https://love2d.org/forums/viewtopic.php?t=11757) and I don't know how to continue (because I don't know how to install `lua-devel` package in a flatpak build):

```
checking for library containing lua_pcall... no
configure: error: Can't LÖVE without Lua
```

Also, since all other games from http://stabyourself.net/ are LÖVE-based, free and fun, maybe this should be built another way to let the common parts among them to be easy to reuse.

I know it still does not have icons, desktop definition, etc. I just wanted to build it properly before digging into those details.

Could you please help me know how to build and set this as WIP in the mean time? Thanks!

### CODE REVIEW

File: net.stabyourself.nottetris2.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+id: net.stabyourself.nottetris2
+runtime: org.freedesktop.Platform
+runtime-version: 1.6
+sdk: org.freedesktop.Sdk
+command: love 'Not Tetris 2.love'
+finish-args:
+  - --socket=x11
```
Comment (nedrichards) [inline net.stabyourself.nottetris2.yml]: add --share=ipc as well if you use x11.

**Comment by nedrichards** [comment]
bot, build net.stabyourself.nottetris2

**Comment by TingPing** [comment]
> I don't know how to continue (because I don't know how to install lua-devel package in a flatpak build):

So `shared-modules/lua5.3/lua-5.3.2.json` certainly contains the headers. The question then becomes how does the LOVE build system look for these?

**Comment by barthalion** [comment]
Version of LOVE you're using requires either luajit 5.1 or lua 5.1, not 5.3.

**Comment by barthalion** [comment]
Here's snippet from gnomempv: https://github.com/flathub/io.github.GnomeMpv/blob/master/org.luajit.Luajit.json

**Comment by nedrichards** [comment]
bot, build net.stabyourself.nottetris2

**Comment by Yajo** [comment]
I think it still doesn't work, it seems LÖVE needs quite a few dependencies...

**Comment by barthalion** [comment]
https://flathub.org/builds/#/builders/27/builds/200

Failed because `shared-modules` submodule uses ssh. Switch it to https and should get to the actual building.

**Comment by barthalion** [comment]
bot, build net.stabyourself.nottetris2

**Comment by barthalion** [comment]
Please include libdevil module directly (only `shared-modules` is allowed to be a submodule). It also needs autoreconf for aarch64 (look at udev from shared-modules for an example). Also you need to build libpng12 before it (see org.libretro.RetroArch).

File: .gitmodules [hunk 2]
```diff
@@ -1,3 +1,6 @@
 [submodule "shared-modules"]
 	path = shared-modules
-	url = https://github.com/flathub/shared-modules.git
+	url = git@github.com:flathub/shared-modules.git
+[submodule "flatpak-modules"]
+	path = flatpak-modules
+	url = git@github.com:casept/flatpak-modules.git
```
Comment (TingPing) [inline .gitmodules]: Keep all sources in repos we control.

**Comment by barthalion** [comment]
Pushed few fixes. LOVE fails to find SDL:

```
In file included from modules/timer/sdl/wrap_Timer.h:26:0,
                 from modules/timer/sdl/wrap_Timer.cpp:22:
modules/timer/sdl/Timer.h:25:17: fatal error: SDL.h: No such file or directory
 #include <SDL.h>
                 ^
compilation terminated.
```

Also it needs some cleanups after it builds correctly.

**Comment by Yajo** [comment]
It's failing to me at this point, did you forget to commit something?:

    Failed to download sources: module libdevil: Can't find file at libdevil-include_IL-il.h.patch

**Comment by barthalion** [comment]
Yeah, just pushed missing files.

**Comment by nedrichards** [comment]
bot, build net.stabyourself.nottetris2

**Comment by barthalion** [comment]
bot, build net.stabyourself.nottetris2

**Comment by nedrichards** [comment]
bot, build net.stabyourself.nottetris2

File: net.stabyourself.nottetris2.yml [hunk 3]
```diff
@@ -2,7 +2,9 @@ id: net.stabyourself.nottetris2
 runtime: org.freedesktop.Platform
```
Comment (barthalion) [inline net.stabyourself.nottetris2.yml]: You still need appdata file, an icon and desktop launcher.

File: net.stabyourself.nottetris2.yml [hunk 4]
```diff
@@ -2,7 +2,9 @@ id: net.stabyourself.nottetris2
...
 runtime-version: 1.6
 sdk: org.freedesktop.Sdk
-command: love 'Not Tetris 2.love'
+command:
+  - love
+  - /files/nottetris2.love
 finish-args:
   - --socket=x11
```
Comment (barthalion) [inline net.stabyourself.nottetris2.yml]: Also --socket=pulsaudio for sound.

File: net.stabyourself.nottetris2.yml [hunk 5]
```diff
@@ -77,7 +79,12 @@ modules:
         path: love-sdl-headers-path.patch
 
   - name: nottetris2
+    buildsystem: simple
+    build-commands:
+      - unzip nottetris2-linux.zip
+      - install -Dm644 'Not Tetris 2.love' /files/nottetris2.love
```
Comment (barthalion) [inline net.stabyourself.nottetris2.yml]: /app/files

File: net.stabyourself.nottetris2.yml [hunk 6]
```diff
@@ -77,7 +79,12 @@ modules:
...
     sources:
-      - type: archive
```
Comment (barthalion) [inline net.stabyourself.nottetris2.yml]: Will be solved via https://github.com/flatpak/flatpak-builder/pull/166 but it's fine as workaround for now.

File: net.stabyourself.nottetris2.yml [hunk 7]
```diff
@@ -0,0 +1,90 @@
+id: net.stabyourself.nottetris2
+runtime: org.freedesktop.Platform
+runtime-version: 1.6
+sdk: org.freedesktop.Sdk
+command:
+  - love
```
Comment (barthalion) [inline net.stabyourself.nottetris2.yml]: It won't work this way. The best would be to create a wrapper script.

**Comment by nedrichards** [comment]
bot, build net.stabyourself.nottetris2

**Comment by Yajo** [comment]
From my 2nd-to-none experience, I'd say this is finished now.

However, I'm getting this error at build:

```
Exporting net.stabyourself.nottetris2 to repo
error: min-free-space-percent '3%' would be exceeded
Export failed: Child process exited with code 1
```

Any further help? Thanks for your patience :sweat_smile:

**Comment by barthalion** [comment]
Does it start for you? It fails with this message here:

```
Could not set video mode: X11 driver not configured with OpenGL
Error: [string "boot.lua"]:1: Could not set screen mode
stack traceback:
	[C]: in function 'assert'
	[string "boot.lua"]:1: in function <[string "boot.lua"]:1>
	[C]: in function 'xpcall'
```

**Comment by Yajo** [comment]
Yikes! I managed to build it and I'm having the exact same problem when running it...

After googling awhile, all points to the same direction: SDL built without OpenGL support. See:

- https://www.phoronix.com/forums/forum/linux-graphics-x-org-drivers/amd-linux/7386-help-x11-driver-not-configured-for-opengl?p=100122#post100122
- https://love2d.org/forums/viewtopic.php?t=18971#p83712

However I have completely no idea of how to enable that. I've read about setting the `SDL_VIDEO_GL_DRIVER`. Do any of you have any clues about that? :thinking:

**Comment by nedrichards** [comment]
bot, build net.stabyourself.nottetris2

File: net.stabyourself.nottetris2.yml [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+  - name: physfs
+    buildsystem: cmake-ninja
+    config-opts:
+      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
+    sources:
+      - type: archive
+        url: http://icculus.org/physfs/downloads/physfs-3.0.1.tar.bz2
+        sha256: b77b9f853168d9636a44f75fca372b363106f52d789d18a2f776397bf117f2f1
+
+  - name: libdevil
```
Comment (nedrichards) [inline net.stabyourself.nottetris2.yml]: https://flathub.org/builds/#/builders/40/builds/356

this doesn't compile on aarch64. looks like it needs to be reconfigured in a manner similar to SDL here https://github.com/flathub/shared-modules/blob/08e9b9eead7a95139be5d1fbc24bea730fb78f7a/SDL/SDL_net-1.2.8.json

Comment (Yajo) [reply net.stabyourself.nottetris2.yml]: Not sure if I understood you; see 299d8a9. Still I cannot execute it.

**Comment by nedrichards** [comment]
bot, build net.stabyourself.nottetris2

**Comment by barthalion** [comment]
bot, build net.stabyourself.nottetris2

**Comment by barthalion** [comment]
bot, build net.stabyourself.nottetris2

**Comment by barthalion** [comment]
bot, build net.stabyourself.nottetris2

**Comment by barthalion** [comment]
Of course now it fails on the latest runtime. I think it's lost cause…

**Comment by Yajo** [comment]
> I think it's lost cause…

:broken_heart: Oh shame, I love this game so much... I thought flatpak could save it for the future generations... :cry: 

I guess we can close it because I have no idea on how to fix this now, and it seems you guys neither. I sent an email to the original devs team; let's see if they want to help us...

**Comment by barthalion** [comment]
Nah, giving up. I'd need to do too many changes to love itself to make it build. Sorry. :(

**Comment by Yajo** [comment]
Somebody from the love community volunteered to migrate this game's code to newer love version. Let's see if it happens and then I'll reopen. Thank you very much!

**Comment by barthalion** [comment]
Hey @Yajo, if you apply this patch to your PR, it will build and work fine:

```
commit 4f2f64b27f188a20e9959802e6664ca398f1ecfa
Author: Bartłomiej Piotrowski <b@bpiotrowski.pl>
Date:   Fri May 7 13:16:19 2021 +0200

    Bump

diff --git a/love-buildfix.patch b/love-buildfix.patch
new file mode 100644
index 0000000..8654f72
--- /dev/null
+++ b/love-buildfix.patch
@@ -0,0 +1,108 @@
+diff --git a/src/Makefile.am b/src/Makefile.am
+index 53ef0a6..0663211 100644
+--- a/src/Makefile.am
++++ b/src/Makefile.am
+@@ -1,4 +1,4 @@
+-AM_CPPFLAGS = -I. -I./modules -I/usr/include/AL -I/usr/include/freetype2  $(INCLUDE_LUA) -I/usr/include/SDL $(FILE_OFFSET)
++AM_CPPFLAGS = -I. -I./modules -I/usr/include/AL -I/usr/include/freetype2  $(INCLUDE_LUA) -I/app/include/SDL $(FILE_OFFSET)
+ AUTOMAKE_OPTIONS = subdir-objects
+ DEFAULT_INCLUDES =
+ SUBDIRS =
+@@ -366,4 +366,4 @@ love_SOURCES = \
+ ./common/Exception.h \
+ ./common/Exception.cpp \
+ ./common/EnumMap.h \
+-./common/Data.h
+\ No newline at end of file
++./common/Data.h
+diff --git a/src/Makefile.in b/src/Makefile.in
+index 391bb6f..f95dd99 100644
+--- a/src/Makefile.in
++++ b/src/Makefile.in
+@@ -393,7 +393,7 @@ target_alias = @target_alias@
+ top_build_prefix = @top_build_prefix@
+ top_builddir = @top_builddir@
+ top_srcdir = @top_srcdir@
+-AM_CPPFLAGS = -I. -I./modules -I/usr/include/AL -I/usr/include/freetype2  $(INCLUDE_LUA) -I/usr/include/SDL $(FILE_OFFSET)
++AM_CPPFLAGS = -I. -I./modules -I/usr/include/AL -I/usr/include/freetype2  $(INCLUDE_LUA) -I/app/include/SDL $(FILE_OFFSET)
+ AUTOMAKE_OPTIONS = subdir-objects
+ DEFAULT_INCLUDES = 
+ SUBDIRS = 
+diff --git a/src/modules/font/freetype/Font.h b/src/modules/font/freetype/Font.h
+index a98778f..d8ad364 100644
+--- a/src/modules/font/freetype/Font.h
++++ b/src/modules/font/freetype/Font.h
+@@ -26,14 +26,14 @@
+ 
+ // FreeType2
+ #ifdef LOVE_MACOSX
+-#include <freetype/ft2build.h>
++#include <freetype2/freetype/ft2build.h>
+ #else
+ #include <ft2build.h>
+ #endif
+-#include <freetype/freetype.h>
+-#include <freetype/ftglyph.h>
+-#include <freetype/ftoutln.h>
+-#include <freetype/fttrigon.h>
++#include <freetype2/freetype/freetype.h>
++#include <freetype2/freetype/ftglyph.h>
++#include <freetype2/freetype/ftoutln.h>
++#include <freetype2/freetype/fttrigon.h>
+ 
+ namespace love
+ {
+diff --git a/src/modules/font/freetype/TrueTypeRasterizer.h b/src/modules/font/freetype/TrueTypeRasterizer.h
+index 6fc13db..7393bc2 100644
+--- a/src/modules/font/freetype/TrueTypeRasterizer.h
++++ b/src/modules/font/freetype/TrueTypeRasterizer.h
+@@ -27,10 +27,10 @@
+ 
+ // TrueType2
+ #include <ft2build.h>
+-#include <freetype/freetype.h>
+-#include <freetype/ftglyph.h>
+-#include <freetype/ftoutln.h>
+-#include <freetype/fttrigon.h>
++#include <freetype2/freetype/freetype.h>
++#include <freetype2/freetype/ftglyph.h>
++#include <freetype2/freetype/ftoutln.h>
++#include <freetype2/freetype/fttrigon.h>
+ 
+ namespace love
+ {
+diff --git a/src/modules/graphics/opengl/Framebuffer.cpp b/src/modules/graphics/opengl/Framebuffer.cpp
+index 93575d7..ea36807 100644
+--- a/src/modules/graphics/opengl/Framebuffer.cpp
++++ b/src/modules/graphics/opengl/Framebuffer.cpp
+@@ -1,3 +1,5 @@
++#define GL_GLEXT_PROTOTYPES
++
+ #include "Framebuffer.h"
+ #include <common/Matrix.h>
+ 
+diff --git a/src/modules/graphics/opengl/GLee.h b/src/modules/graphics/opengl/GLee.h
+index 34ca16b..dd3aa2c 100644
+--- a/src/modules/graphics/opengl/GLee.h
++++ b/src/modules/graphics/opengl/GLee.h
+@@ -67,6 +67,7 @@
+ 	#define GLX_GLXEXT_PROTOTYPES
+ 	#include <GL/gl.h>
+ 	#include <GL/glx.h>
++typedef XID GLXContextID;
+ #endif
+ 
+ #ifndef APIENTRY
+diff --git a/src/modules/graphics/opengl/SpriteBatch.cpp b/src/modules/graphics/opengl/SpriteBatch.cpp
+index ca5762f..7a2e57a 100644
+--- a/src/modules/graphics/opengl/SpriteBatch.cpp
++++ b/src/modules/graphics/opengl/SpriteBatch.cpp
+@@ -18,6 +18,8 @@
+ * 3. This notice may not be removed or altered from any source distribution.
+ **/
+ 
++#define GL_GLEXT_PROTOTYPES
++
+ #include "SpriteBatch.h"
+ 
+ // STD
diff --git a/love-glext-prototypes.patch b/love-glext-prototypes.patch
deleted file mode 100644
index 514de3d..0000000
--- a/love-glext-prototypes.patch
+++ /dev/null
@@ -1,21 +0,0 @@
-diff --unified --recursive --text love-HEAD.orig/src/modules/graphics/opengl/Framebuffer.cpp love-HEAD.new/src/modules/graphics/opengl/Framebuffer.cpp
---- love-HEAD.orig/src/modules/graphics/opengl/Framebuffer.cpp	2019-03-14 12:46:55.032982224 -0400
-+++ love-HEAD.new/src/modules/graphics/opengl/Framebuffer.cpp	2019-03-14 12:47:22.356175299 -0400
-@@ -1,3 +1,5 @@
-+#define GL_GLEXT_PROTOTYPES
-+
- #include "Framebuffer.h"
- #include <common/Matrix.h>
-
-diff --unified --recursive --text love-HEAD.orig/src/modules/graphics/opengl/SpriteBatch.cpp love-HEAD.new/src/modules/graphics/opengl/SpriteBatch.cpp
---- love-HEAD.orig/src/modules/graphics/opengl/SpriteBatch.cpp	2019-03-14 12:46:55.032982224 -0400
-+++ love-HEAD.new/src/modules/graphics/opengl/SpriteBatch.cpp	2019-03-14 12:47:33.346119890 -0400
-@@ -18,6 +18,8 @@
- * 3. This notice may not be removed or altered from any source distribution.
- **/
-
-+#define GL_GLEXT_PROTOTYPES
-+
- #include "SpriteBatch.h"
-
- // STD
diff --git a/love-sdl-headers-path.patch b/love-sdl-headers-path.patch
deleted file mode 100644
index 764a6e8..0000000
--- a/love-sdl-headers-path.patch
+++ /dev/null
@@ -1,30 +0,0 @@
-diff --git a/src/Makefile.am b/src/Makefile.am
-index 53ef0a6..0663211 100644
---- a/src/Makefile.am
-+++ b/src/Makefile.am
-@@ -1,4 +1,4 @@
--AM_CPPFLAGS = -I. -I./modules -I/usr/include/AL -I/usr/include/freetype2  $(INCLUDE_LUA) -I/usr/include/SDL $(FILE_OFFSET)
-+AM_CPPFLAGS = -I. -I./modules -I/usr/include/AL -I/usr/include/freetype2  $(INCLUDE_LUA) -I/app/include/SDL $(FILE_OFFSET)
- AUTOMAKE_OPTIONS = subdir-objects
- DEFAULT_INCLUDES =
- SUBDIRS =
-@@ -366,4 +366,4 @@ love_SOURCES = \
- ./common/Exception.h \
- ./common/Exception.cpp \
- ./common/EnumMap.h \
--./common/Data.h
-\ No newline at end of file
-+./common/Data.h
-diff --git a/src/Makefile.in b/src/Makefile.in
-index 391bb6f..f95dd99 100644
---- a/src/Makefile.in
-+++ b/src/Makefile.in
-@@ -393,7 +393,7 @@ target_alias = @target_alias@
- top_build_prefix = @top_build_prefix@
- top_builddir = @top_builddir@
- top_srcdir = @top_srcdir@
--AM_CPPFLAGS = -I. -I./modules -I/usr/include/AL -I/usr/include/freetype2  $(INCLUDE_LUA) -I/usr/include/SDL $(FILE_OFFSET)
-+AM_CPPFLAGS = -I. -I./modules -I/usr/include/AL -I/usr/include/freetype2  $(INCLUDE_LUA) -I/app/include/SDL $(FILE_OFFSET)
- AUTOMAKE_OPTIONS = subdir-objects
- DEFAULT_INCLUDES = 
- SUBDIRS = 
diff --git a/net.stabyourself.nottetris2.yml b/net.stabyourself.nottetris2.yml
index 90d444f..a3b0cbe 100644
--- a/net.stabyourself.nottetris2.yml
+++ b/net.stabyourself.nottetris2.yml
@@ -1,6 +1,6 @@
 id: net.stabyourself.nottetris2
 runtime: org.freedesktop.Platform
-runtime-version: "18.08"
+runtime-version: "20.08"
 sdk: org.freedesktop.Sdk
 command: /app/bin/nottetris2.sh
 finish-args:
@@ -8,9 +8,10 @@ finish-args:
   - --share=network
   - --socket=pulseaudio
   - --socket=x11
+  - --device=dri
 modules:
+  - shared-modules/glu/glu-9.json
   - shared-modules/SDL/SDL-1.2.15.json
-  - shared-modules/glu/glu-9.0.0.json
 
   # TODO: use luajit for x86 and x86_64
   - name: lua
@@ -22,12 +23,6 @@ modules:
       - type: patch
         path: lua.patch
 
-  - name: libpng
-    sources:
-      - type: archive
-        url: https://downloads.sourceforge.net/sourceforge/libpng/libpng-1.6.34.tar.xz
-        sha256: 2f1e960d92ce3b3abd03d06dfec9637dfbd22febf107a536b44f7a47c60659f6
-
   - name: libmng
     sources:
       - type: archive
@@ -83,17 +78,13 @@ modules:
         sha256: a57adcb0cbdc390a9bd8e2fe477bc175799b9ffd3486e01f859a36bf27f7f268
 
       - type: patch
-        path: love-sdl-headers-path.patch
-
-      - type: patch
-        path: love-glext-prototypes.patch
+        path: love-buildfix.patch
 
       - type: script
         dest-filename: autogen.sh
         commands:
           - AUTOMAKE="automake --foreign" autoreconf -vfi
 
-
   - name: nottetris2
     buildsystem: simple
     build-commands:
diff --git a/shared-modules b/shared-modules
index 27e7d67..fd4d383 160000
--- a/shared-modules
+++ b/shared-modules
@@ -1 +1 @@
-Subproject commit 27e7d67b60a30ffd2b807a4f3eb3f65b0e9df953
+Subproject commit fd4d38328ccb078b88ad4a891807e593ae8de806
```

For your convenience, it's also at https://github.com/flathub/flathub/commit/4f2f64b27f188a20e9959802e6664ca398f1ecfa

**Comment by barthalion** [comment]
It also needs `--device=dri`.

**Comment by Yajo** [comment]
Thanks! I cherry-picked it here.

bot, build net.stabyourself.nottetris2

**Comment by barthalion** [comment]
Appdata misses <releases> section.

**Comment by Yajo** [comment]
bot, build net.stabyourself.nottetris2

It works!

**Comment by Yajo** [comment]
Oh I can't believe it! Soo many years without this game, and now it works again! 🙈 🎉 

@Stabyourself and @barthalion, do you want to maintain or share maintenance of the flathub package?

**Pull request closed without merging**